### PR TITLE
Convert functions to use code points

### PR DESCRIPTION
**Pull request by MonoidMusician**
This is more accurate because `CodePoint` represents **any** unicode character, whereas `Char` only works with non-astral characters, due to its JavaScript representation. Right?

I'll probably submit a PR to `-strings` to add an equivalent of `charPoint` directly to Data.String.CodePoints, which will clean this up a bit.

### CODE REVIEW

**Comment by cdepillabout** [comment]
@MonoidMusician Thanks for this PR!

To be honest, I'm not super knowledgable about unicode.  Would you be able to link me to something explaining the difference between Javascript `Char`s and actual unicode characters?

Before I merge this, could you do the following things:

- update the tests with unit tests that are not possible to write when using `Char`, but are now possible when using `CodePoint`
- update the examples in the docoments for each function (for instance the `isPunctuation` function: https://github.com/purescript-contrib/purescript-unicode/pull/15/files#diff-a8371e327cfff9b134c465f53ecdd381L407)
- create a `Changelog.md`, add it to `package.json` (or `bower.json` or whereever it needs to go), and add an explanation about changing `Char` to `CodePoint`, linking back to this PR

Also, [purescript-parsing](https://github.com/purescript-contrib/purescript-parsing) depends on purescript-unicode.  After this is merged, would you be willing to send a small PR to purescript-parsing updating their uses of purescript-unicode?

**Comment by cdepillabout** [comment]
@michaelficarra, would it be possible for you to take a quick glance at this PR and make sure everything looks sane?

It looks like you're the author of `Data.String.CodePoint`, so I was thinking you might be the most qualified to tell whether this use of it is good.

In case you haven't see `purescript-unicode` before, it is basically a straight copy of Haskell's [`Data.Char`](https://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Char.html)  module.

**Comment by cdepillabout** [comment]
Also, I wonder if it makes sense to change the module name from `Data.Char.Unicode`, to something like `Data.CodePoint.Unicode`.

File: src/Data/Char/Unicode.purs
Comment (michaelficarra) [inline src/Data/Char/Unicode.purs]: `toUpper` can't be `CodePoint -> CodePoint`, it must be `CodePoint -> String`. But maybe that's something we should address separately, since I see it's currently broken anyway.

**Comment by michaelficarra** [comment]
I don't like the use of `unsafePartial`, but yes, this looks like the right direction. Also, I agree that something like `charPoint` should be built in.

File: src/Data/Char/Unicode.purs
Comment (MonoidMusician) [reply src/Data/Char/Unicode.purs]: Okay ... Is this because a some precomposed characters don't have direct upper case equivalents and would need to be decomposed?

Comment (michaelficarra) [reply src/Data/Char/Unicode.purs]: Possibly, but also see https://unicode.org/faq/casemap_charprop.html#11

**Comment by MonoidMusician** [comment]
@cdepillabout Thanks for the suggestions!

According the the documentation for Prim.Char [1], it represents one UTF-16 code unit, i.e. all _non-astral_ code points. But CodePoint is a newtype around Int so it can represent any Unicode value.

``Int `superset` CodePoint `superset` Char`` I believe

I should make a PR for `-strings` to make most of those changes ... maybe `codePoint{To,From}Char` would be a good name instead of `charPoint`?

[1] https://pursuit.purescript.org/builtins/docs/Prim#t:Char

File: src/Data/Char/Unicode.purs
Comment (MonoidMusician) [reply src/Data/Char/Unicode.purs]: Okay, cool ... I'm not familiar with the representation of the rules, so I wouldn't know how to fix that. Also, would we want to return `String` or `Array CodePoint`? (maybe the latter for the the `Internal` module?)

Comment (michaelficarra) [reply src/Data/Char/Unicode.purs]: `Array CodePoint` sounds good.

Comment (MonoidMusician) [reply src/Data/Char/Unicode.purs]: currently it looks like `toUpper` does not change the eszett.

Comment (cdepillabout) [reply src/Data/Char/Unicode.purs]: @MonoidMusician could you also add some documentation to `toUpper` explaining why the type is `CodePoint -> Array CodePoint`?

Also, I'm wondering if it would be helpful to have a function like `unsafeToUpper :: CodePoint -> CodePoint` for people who want to easily map over a list of codepoints, without worrying about being 100% correct.  Its use should be discouraged of course.

**Comment by cdepillabout** [comment]
> Also, I wonder if it makes sense to change the module name from `Data.Char.Unicode`, to something like `Data.CodePoint.Unicode`.

After thinking about it some more, I think this might be a good change to make.

It would make it possible for us to go back and re-add a `Data.Char.Unicode` module that works on `Char` instead of `CodePoint`.  That would make it easy for people to use without having to figure out the relationship between `Char` and `CodePoint`.  (I'm assuming that most users

The documentation for `Data.Char.Unicode` should point out its deficiencies, and also point users to the more-correct `Data.CodePoint.Unicode` module.

**Comment by michaelficarra** [comment]
Why even have the `Data.Char` one? We have code unit based functions in `-strings` only for legacy and are considering moving around the modules to make `CodePoint` the default.

**Comment by cdepillabout** [comment]
@michaelficarra 

> We have code unit based functions in -strings only for legacy and are considering moving around the modules to make CodePoint the default.

Does "code unit" mean PureScript's `Char`?

I was thinking that there would be a lot of [beginner?] PureScript programmers that are using `Char` (since it is explained in the PureScript book and available in `Prim`).  Many of them may want functions like `toUpper`, `toLower`, `isSymbol` etc they can easily use, without having to figure out the `CodePoint` type.  That's why we should also provide a `Data.Char.Unicode` module in addition to (my proposed) `Data.CodePoint.Unicode`.

However, there will certainly be people that want to handle unicode correctly.  They can use the `Data.CodePoint.Unicode` module.  The `Data.Char.Unicode` should specifically recommend using `Data.CodePoint.Unicode`.

-------------------------------------------

However, if the rest of the PureScript community is moving to completely use `CodePoint` instead of `Char`, then I think that you are correct.  We don't really need the `Data.Char.Unicode` module.

**Comment by MonoidMusician** [comment]
I will update this PR once this is merged and released: https://github.com/purescript/purescript-strings/pull/92

It looks like the mapping of “ß” `00DF` to “SS” `0073 0073` is not specified in `UnicodeData.txt` but rather [`CaseFolding.txt`](http://www.unicode.org/Public/10.0.0/ucd/CaseFolding.txt). Should we parse that as well and use that for case data? Like you said, `CodePoint -> Array CodePoint` makes sense here, and we can also expose `String -> String` by using `concatMap` with the appropriate conversions.

```
# UnicodeData.txt specifies:
00DF;LATIN SMALL LETTER SHARP S;Ll;0;L;;;;;N;;;;;
# CaseFolding.txt specifies:
00DF; F; 0073 0073; # LATIN SMALL LETTER SHARP S
```

I've started my own standard parsing library so if I could probably attempt to convert our parser here to PS from awk: https://github.com/MonoidMusician/purescript-uievents-key/blob/master/test/Generate.purs

Also should we generate it from the latest 10.0.0 instead of 6.0.0 linked in the README? I'm not sure how significant the changes are (more emoji anyone?? 😄)

And my last point: I renamed it to `Data.CodePoints.Unicode` but the strings package uses `Data.String.CodePoints` ... should I go to `Data.String.CodePoints.Unicode` to be consistent with that? I personally don't think that keeping `Char` around makes much sense, especially since we will be able to easily upcast that once the PR is merged.

Thanks for the feedback from both of you!

**Comment by michaelficarra** [comment]
> Should we parse that as well and use that for case data?

Yes, if you want to provide case mapping functions, you should base them on that table.

> Also should we generate it from the latest 10.0.0 instead of 6.0.0 linked in the README?

Yes, but stay consistent within major releases of purescript-unicode.

> should I go to `Data.String.CodePoints.Unicode` to be consistent with that?

No, see https://github.com/purescript/purescript-strings/issues/95.

**Comment by cdepillabout** [comment]
It looks like there has been some discussion about moving to CodePoints on https://github.com/purescript/purescript-strings/issues/95, so it seems like it might be a good time to revisit this issue!

**Comment by kritzcreek** [comment]
I'm going to need a response to this "quickly" or I'll make a 0.12 release with the old API and we'll need to make another breaking release to incorporate this PR.

**Comment by cdepillabout** [comment]
@MonoidMusician Does this PR need to be updated?  Or can it be merged in as-is?

**Comment by MonoidMusician** [comment]
I am starting to work on updating it :) Special casing might be a separate PR though ... maybe we should keep `toUpper`/`toLower` as-is (or rename now), and add new functions for [Special Casing](https://unicode.org/Public/UNIDATA/SpecialCasing.txt)?

**Comment by kritzcreek** [comment]
@MonoidMusician For anything that you do, make sure you do it against the compiler/0.12 branch. This package-set should have all the dependencies you need: https://github.com/kRITZCREEK/package-sets/tree/test-core

EDIT:

fromCharCode returns a Maybe now, but since this library is pretty low-level you might want to redefine a non-Maybe version with the FFI and use that whenever you're sure you'll stay within bounds.

**Comment by MonoidMusician** [comment]
Hm, the dev dependency `purescript-spec` isn't updated ...

**Comment by MonoidMusician** [comment]
I had to fix up the testing libraries locally, but the tests pass here now. @kRITZCREEK suggested using `purescript-assert` so we don't have to "wait" on spec being updated ...

**Comment by hdgarrood** [comment]
I'm interested in helping move this forward; does anyone have suggestions for how I could do that?

**Comment by thomashoneyman** [comment]
@MonoidMusician We're making breaking changes for libraries for the upcoming 0.14 release. I know it's been a long, long time since you've looked at this, but I'm curious if you have any advice or suggestions for folks who would like to move this along and get it in for the upcoming breaking release. Thanks!

**Comment by MonoidMusician** [comment]
I got around to this again! I added a script to scrape data from both CaseFolding.txt and SpecialCasing.txt – it turns out that the former only specifies canonical casings (and has full and simple variants), while the latter is the full lower/title/upper mappings that we were looking for.

Are we happy with the types and naming?

I need to think more about what `toTitle :: String -> String` will look like …

Should also update to a later version of the standard.

**Comment by thomashoneyman** [comment]
@MonoidMusician CI was failing because of the hash (as you noticed) as well as an old release candidate of the compiler. I've bumped that for you 😄

**Comment by thomashoneyman** [comment]
@MonoidMusician CI is failing because you've removed all JavaScript files under `src`. If there will no longer be FFI in this project, then you can:

- delete the `package.json` file
- delete the `.eslintrc.json` file
- replace the `.github/workflows/ci.yml` file with the standard one for libraries not using JavaScript:
  https://github.com/purescript-contrib/purescript-machines/blob/main/.github/workflows/ci.yml

That will return this library to the standard setup used in `contrib` libraries when they don't have FFI.

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [inline src/Data/String/Unicode.purs]: If `conv` stands for `convert`, can we use the full name?

```suggestion
convert :: (CodePoint -> CodePoint) -> String -> String
convert f = toCodePointArray >>> map f >>> fromCodePointArray

convertFull :: (CodePoint -> Array CodePoint) -> String -> String
convertFull f = toCodePointArray >>> bindFlipped f >>> fromCodePointArray
```

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [inline src/Data/String/Unicode.purs]: Is this meant to be commented out? Should it be removed if so?

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [inline src/Data/String/Unicode.purs]: Could we add a brief note describing in what situations this may be faster than `caseFold`?

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [inline src/Data/String/Unicode.purs:45]: Should this perhaps be `caselessEq` as this function checks equality? If `match` is the more correct term then feel free to resolve this without changes.

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [inline src/Data/String/Unicode.purs]:
```suggestion
-- | Because this matches on more rules, it may be slower than `Data.CodePoint.Unicode.toLower`, but it
```

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [inline src/Data/String/Unicode.purs]:
```suggestion
-- | Because this matches on more rules, it may be slower than `Data.CodePoint.Unicode.toUpper`, but it
```

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [inline src/Data/String/Unicode.purs]: These don't seem to be used anywhere. Can they be removed?

File: unicode-version
Comment (thomashoneyman) [inline unicode-version:1]: Could we add a line to the README file that specifies we're supporting Unicode 13.0.0? We could include the link:

```md
This version of `unicode` supports the [Unicode standard, version 13.0.0](https://www.unicode.org/versions/Unicode13.0.0).
```

File: src/Data/CodePoint/Unicode.purs
Comment (thomashoneyman) [inline src/Data/CodePoint/Unicode.purs]: Can we preserve this comment?

```suggestion
  , digitToInt -- Deprecated
```

File: src/Data/CodePoint/Unicode.purs
Comment (thomashoneyman) [inline src/Data/CodePoint/Unicode.purs]: Can we preserve this comment?

```suggestion
  , isDigit -- Deprecated
```

File: .github/workflows/ci.yml
Comment (thomashoneyman) [inline .github/workflows/ci.yml:29]:
```suggestion

      - name: Install dependencies
```

Comment (thomashoneyman) [reply .github/workflows/ci.yml:29]: (Just a minor suggestion to preserve the newline)

File: src/Data/CodePoint/Unicode/Internal.purs
Comment (thomashoneyman) [inline src/Data/CodePoint/Unicode/Internal.purs:3]: I don't have a good way to verify this file other than seeing that the tests still pass. From a scan through it looks fine to me. If you have any suggestions on verifying this, @MonoidMusician, I'm all ears!

File: fullcase.js
Comment (thomashoneyman) [inline fullcase.js:72]: Could we add documentation comments to this and the declarations throughout the file, as this is a public module? Simple ones are fine, but they can help guide maintainers in the future as well.

**Review by thomashoneyman — CHANGES_REQUESTED** [review]
This is really impressive, @MonoidMusician, and thank you for your work. I only have minor comments to make; on the whole it's good to see this modernized and ready for Unicode 13.0.0, as well as on a sounder footing with the switch to code points.

Admittedly, I'm not particularly well-versed in the subtleties of unicode, so perhaps if @michaelficarra has time for another look that would be helpful as well.

File: src/Data/CodePoint/Unicode/Internal.purs
Comment (thomashoneyman) [reply src/Data/CodePoint/Unicode/Internal.purs:3]: Also, we have some documentation that describes generating these files anew:

https://github.com/purescript-contrib/purescript-unicode/tree/main/docs#generating-internal-modules

You've made some changes to how these are generated; if there's any information you think would be helpful to maintainers working on this in the future, please consider adding it to that documentation file as well. Thanks!

File: src/Data/String/Unicode.purs
Comment (MonoidMusician) [reply src/Data/String/Unicode.purs]: Oh good catch, I actually meant `toUpperSimple` here. (Forgot to update the comments when I switched names.)

File: src/Data/CodePoint/Unicode/Internal.purs
Comment (MonoidMusician) [reply src/Data/CodePoint/Unicode/Internal.purs:3]: Thanks, I missed that it was moved to there, will update.

File: fullcase.js
Comment (MonoidMusician) [reply fullcase.js:72]: I moved it to a new Internal folder – users should not use these functions on `Int`s but instead the ones on `CodePoint`s.

File: src/Data/String/Unicode.purs
Comment (MonoidMusician) [reply src/Data/String/Unicode.purs:45]: The Unicode standards use the term caseless matching, but `caselessEq` might be more clear. I don't have a strong opinion.

File: src/Data/String/Unicode.purs
Comment (MonoidMusician) [reply src/Data/String/Unicode.purs]: Yeah, I'll just remove it. There's no clear way to divide strings into words to do proper titlecasing – if users need that they can implement an appropriate algorithm themselves using the functions in `Data.CodePoint.Unicode`.

**Comment by MonoidMusician** [comment]
Okay, thanks for the feedback! I think I addressed what was addressable.

File: packages.dhall
Comment (thomashoneyman) [inline packages.dhall]:
```suggestion
      https://raw.githubusercontent.com/purescript/package-sets/prepare-0.14/src/packages.dhall
```

File: src/Data/String/Unicode.purs
Comment (thomashoneyman) [reply src/Data/String/Unicode.purs:45]: OK, I think it's reasonable to stay with the standard. Thanks!

**Review by thomashoneyman — APPROVED** [review]
👍 Looks good to me! Thanks for your work! If no other maintainer notices an issue before this must be merged in order to make the 0.14 release, then I will merge this PR. I'm going to leave it open in the meantime.

File: fullcase.js
Comment (michaelficarra) [inline fullcase.js]:
```suggestion
const rhex = a => parseInt(a, 16);
```

Do we expect any of these keys to not be strings?

Comment (michaelficarra) [inline fullcase.js:17]:
```suggestion
const rhex = a => parseInt(a, 16);
```

Do we expect any of these keys to not be strings?

File: fullcase.js
Comment (michaelficarra) [inline fullcase.js:20]: `String.prototype.matchAll` is a very recent feature. What kind of engine support are we targeting with this project? We can accomplish the same thing without it if we need.

**edit:** Since this is a development time script, I guess we can assume a very modern engine.

Comment (MonoidMusician) [reply fullcase.js:20]: This is just for generating code, it's not part of the runtime of the library. So I assume that developers can find a recent enough node.

File: README.md
Comment (michaelficarra) [inline README.md]: I would just link to `unicode-version` and say that the currently supported version can be found in that file. We plan on updating this project following the release of a new Unicode version, right?

File: fullcase.js
Comment (michaelficarra) [inline fullcase.js]: The first line of generated files should be a comment that says it was generated and what script generated it. Additionally, you should mark generated files in the `.gitattributes` so that GitHub suppresses their diffs. See https://github.com/github/linguist#generated-code

**Review by michaelficarra — COMMENTED** [review]
Generally LGTM. What do you think about also exporting the Unicode version as a String?

**Comment by MonoidMusician** [comment]
Okay, I think I addressed your feedback, thanks! I don't think we need to export the Unicode version from the library, I just want to make sure it's available for documentation when someone goes looking for it. Unicode seems to be pretty stable so I don't think it affects most users.

**Comment by MonoidMusician** [comment]
I just noticed that this fixes #28 (I used hexadecimal escapes).

**Comment by thomashoneyman** [comment]
Fantastic work, @MonoidMusician! Thank you, and if anyone does happen to notice an issue we can merge a fix.

**Pull request merged**
